### ISSUE TITLE
withTolerance() but only for doubles and floats

### ISSUE THREAD

**Issue by fprochazka** [lukas-krecan/JsonUnit#629]
Hi,
I have a JSON that I have to pipe through Jackson so that I can prettyprint it. The problem is that Jackson changes the doubles from `7.00` to `7.0`, which I don't like but I don't really care that much... 

problem is that json-unit then (rightfully) complains the values are different... which can be worked around by configuring `.withTolerance(0)`, but that means it also accepts `7`... which makes sense mathematically, but not from types point of view... 

when parsing the JSON, it doesn't really matter if its `7.0` or `7.00`, because both of them will result in float/double value in _most_ programming languages, but `7` vs `7.0` is a problem, because one of them is float/double and one of them is int. 

So I'm proposing to introduce two new configuration methods (I'm actually not sure what names would be good, its just an example)

```java
.withFloatingPointTolerance(BigDecimal)
.withWholeNumberTolerance(BigDecimal)
```

and changing the `.withTolerance()` to

```java
    public Configuration withTolerance(@Nullable BigDecimal tolerance) {
        return withFloatingPointTolerance(tolerance).withWholeNumberTolerance(tolerance);
    }
```

this way, the compatibility is preserved, I can safely assert float/double values mathematically, and int vs float/double strictly, which should help me to prevent changing the value from float/double to int by mistake

**Comment by lukas-krecan** [comment]
Hi, thanks for feedback. I have few questions to make sure I understand:
1. What's the reason you can't use `7.0` in the "expected" part of the assertion
2. What would the `withWholeNumberTolerance` actualy do? It seems strange to me to say `withWholeNumberTolerance(0.03)`, it's the whole numbers, no tolerance should be needed. Moreover, it's not clear to me if you mean whole numbers in the expected or actual JSON. Would it apply both when you compare `7 == 7.0` and `7.0 == 7`?

If I understand you correctly, you want an exact match in mathematical sense, but taking the types into account? Or is there some rounding at play too?

**Comment by fprochazka** [comment]
Sorry for chaotic report, your summary is on point 

> If I understand you correctly, you want an exact match in mathematical sense, but taking the types into account?

**Comment by lukas-krecan** [comment]
I see. I am afraid I will not be able to design configuration covering all the posibilities. But I can add some `NumericalMatcher` interface and let you plug-in your own comparison. Would that work for you?

**Comment by fprochazka** [comment]
I think that would be perfectly sufficient :)

**Comment by lukas-krecan** [comment]
Something line this https://github.com/lukas-krecan/JsonUnit/pull/632/files#diff-642a52fede8473f98c5a7b25f34c6bd68f33160cf22733127f7f7f13f3cb2fc6R777 ?

**Comment by fprochazka** [comment]
Awesome, that looks like it should work! Thank you!

**Comment by lukas-krecan** [comment]
Released as 2.38.0
